### ISSUE TITLE
Simplify entity definition

### ISSUE THREAD

**Issue by fpagnoux** [openfisca/openfisca-core#560]
Suggestions:

- Remove and deprecate subroles
- Assume an entity describes persons if no role is defined.
- Add a `min` property
- Assume a role is unique if it has no plural
- Remove label properties if they are not used

**Comment by MattiSG** [comment]
@benjello We think subroles should be split into flat roles to avoid nesting. Opinion?

**Comment by benjello** [comment]
To be sure that I understand you well, you want to switch from:
 - family:
    - parents:
      - first_parent
      - second_parent
   - children
      - child
      - child 

to 

- family:
  - first_parent
  - second_parent
  - child
  - child

?

**Comment by MattiSG** [comment]
Not really: if Family is an entity and `parent` and `child` are roles, then we'd have the hierarchy you described in the first example. Subroles apparently refine roles, as in “mother” being a possible subrole for “parent”. That adds a lot of complexity, and the suggested change would be to replace these by `parent_mother` in this example.

**Comment by fpagnoux** [comment]
Some context about why subroles even came to exist in the first place:

OpenFisca France has 3 group entities: `famille`, `foyer_fiscal` and `menage`. These entities have  a common structure: one or two people that are "in charge" (typically the parents), and other persons that gravitate around them (typically the children).

One important modelling choice is: do these two people "in charge"  have symmetrical roles? Or is there a clear distinction between them? 

For `menage`, the answer has always been clear for france: the two roles are distinct. I believe this is because INSEE historically loved this idea of "Chef de famille", or "Personne de référence" 👨‍💼 .

For `famille` and `foyer_fiscal`, I'm talking under the control of @benjello but I believe OpenFisca France leant more towards symmetric roles. The way to declare a `famille` has always been something like `{parents: [Alice, Bob], children: [Kevin]}`. Because, _in the model_, there is no reason to distinguish the two parents.

However, practically, indistinguishable parents were challenging in two situations. Both related to projecting values from one group entity to another. This is where subroles came to help.

1. I'm writing a formula for a variable `benefit`, defined for a `famille`. I need to access a variable 
 `code_postal`, defined for a `menage`. But which one of the `famille`'s member `menage` should I use? Having a subrole pointing to _one person_ helped. I could just use: `famille.demandeur.menage('code_postal', period)`, with the convention that the first `parent` of a `famille` has the subrole `demandeur`.

2. I'm writing a formula for a `menage` variable, `disposable_income`, to do some sort of country-level statistics. I'd like to include the `benefits` received by the `famille(s)` inside the `menage`. But I need to be careful not to double count. If two persons are in the same `famille` but in different `menage`, I want to count the `benefits` _only once_. So this is what I do:

```py
benefits_ind = menage.members.famille('benefits', period)
# Only sums for menage members whose role in famille is DEMANDEUR.
# This way, I'm sure that for each famille the benefits are projected on only one menage
benefits_menage = menage.sum(benefits_ind, role = Famille.DEMANDEUR)  
```

So **TL/DR: Modelling pushed for symmetric roles, convenience imposed distinguishable roles. Subroles came as a middle-of-the-road bastard solution that allowed both**

**Comment by fpagnoux** [comment]
Having said that, I do agree that this solution is too complex, a little too tailored-made, and is going to make new users frown eyebrows.

I see two escapes:

- Make France flatten their roles. Too bad for the modeling aspect. The rule would be: _if you want to have several group entities and make projection from one to another, then yes, practically, you need at least one unique role in your entity_.
- Offer ways to solve the two issues mentioned above without subroles.

The first one is actually easy to solve: just allow this syntax: `famille.parent[0].menage('code_postal', period)`. 

The second is more challenging, but I believe we could do something like:

```py
is_parent = menage.members.role_in(Famille) == Famille.PARENT
is_first_of_role = menage.members.rank_by_role_in(Famille) == 0
is_first_parent = is_parent * is_first_of_role
benefits_menage = menage.sum(benefits_ind * is_first_parent)  
```

(Of couse, `is_first_parent` could be extracted to its own formula)

I estimate making this possible to 1 day of work, plus 0.5 for reviews (I may have already done some experiments that shows there is no big trap).

Now the question is, as usual, is it worth it? I guess it depends how much the italic rule in the beginning of this message feel like a too big constraint for those who model tax and benefit systems

**Comment by benjello** [comment]
@fpagnoux : for the sake of completeness we did start with non-symmetric roles since they were the first used historically and very convenient for projections but then for "political correctness" issues and also legislation evolution we added the layer of symmetric roles.

**Comment by MattiSG** [comment]
> 1. I need to access a variable code_postal, defined for a menage. But which one of the famille's member menage should I use?

I don't understand that. If `code_postal` is defined for a `menage`, how does the role of any Person I use to access that `menage` impact in any way the Menage's variables? If it's just a matter of navigating through _any_ member of one group entity to the other, then couldn't we as well offer a `some_member` property that will choose one randomly (not that I believe it would be a good API, but just to check my understanding)?

> 2. I need to be careful not to double count

Feels like both needs are about navigating from one group entity to another. I believe we could create a simple and readable API for that.

Would something like:

```py
familles = menage.entities_sharing_members(Famille)  # returns a vector of arrays
```

make sense?

**Comment by fpagnoux** [comment]
> how does the role of any Person I use to access that menage impact in any way the Menage's variables?

Because the `menage` of the first parent of the `famille` is not necessarily the `menage` of the first child. So taking a random member of the family may, in some case, return a nondeterministic value for the variable `code_postal`.


> `familles = menage.entities_sharing_members(Famille)  # returns a vector of arrays`

Not sure I fully understand... If I understand correctly, `familles` would be a 2D-matrix for which:
- Each columns represent a `menage`
- Within a column, the ith cell represents the `famille` of the ith person of the `menage`

I imagine I could use `familles` to calculate  the corresponding `benefits` 2D-matrix. But then what would I do with that? I can't just sum by column, because there would be a risk of double-counting.

**Comment by MattiSG** [comment]
> But then what would I do with that? I can't just sum by column

Gotcha. Thanks a lot for bearing with me 🙏

In Aotearoa, the notion of “principal caregiver” seems quite similar to “chef de famille”, and is [implemented as a role](https://github.com/ServiceInnovationLab/openfisca-aotearoa/pull/19/files#diff-7b21b7200161098593bdd862d28fe54cR64). We obviously don't have tried this through time, but conceptually this works well.

Now that we've explored the context and demonstrated the difficulty in navigating from one group entity to the other, I'm quite happy to go back to the initial needs and description of this PR.

**Comment by Morendil** [comment]
Do we want to keep this open?

My understanding is that we still have an open question about subroles (recently discussed with @fpagnoux), the other points in this issue do not strike me as significant pain points coming up in recent discussions, so I would be in favor of closing this to focus more specifically on roles/subroles.

Also #857 will, I hope, bring more flexibility to entity definition.

**Comment by Morendil** [comment]
My current thinking on this is that we can do away with subroles **without replacing them with anything** (which was my beef with #789, which added new and confusing syntax).

Here's a typical use of the subrole `DECLARANT_PRINCIPAL` in France:

```Python
class salaire_imposable(Variable):
  entity = Individu
  def formula(individu, period):
    # ...
    rev_microsocial = individu.foyer_fiscal('rev_microsocial', period, options = [DIVIDE])
    rev_microsocial_declarant1 = rev_microsocial * individu.has_role(FoyerFiscal.DECLARANT_PRINCIPAL)
    # ...
```

Now suppose we have the following structure:
```YAML
- individus: [Alice, Bob, Chris]
- foyers_fiscaux:
  - one: [Alice]
  - two: [Bob, Chris]
```

Suppose for simplicity's sake that the `rev_microsocial` of each tax household is 100. We have two tax households, so the total amount of that variable in the model (whatever it is) is 200.

Our problem is that `individu.foyer_fiscal('rev_microsocial')` will "project" this value to each individual and yield the following array: `[100, 100, 100]`. So if we naively sum over this array we are double-counting and losing consistency; the total amount of that variable in the model, according to the `individu.foyer_fiscal projection`, is 300.

The solution has been to require **the formula** (and hence the formula author) to fix the inconsistency, by ensuring that the value is counted only once per tax household. This is what the second line does. Arbitrarily, the subrole `DECLARANT_PRINCIPAL` is used for that purpose, but we could, with equal justification in this particular example, use the other subrole `CONJOINT`.

The choice I'd like to question is putting the burden on formula authors. **The projection is wrong** and that's what we should fix, instead of having dozens of ad-hoc fixes in formulas; and if we fix the projection, nearly all of the need for the subrole disappears.

To fix the projection we need to have `individu.foyer_fiscal('rev_microsocial')` return an array consistent with the total amount in the model: `[100, 100, 0]`. (The amounts for Alice's and Bob's tax households are 100, and since Chris's tax household is the same as Bob's and is already projected, we drop it to 0.) The projection shouldn't "double project", and if it doesn't then the formula cannot double count. In practice, that means that we can remove the second line in the snippet above.
